### PR TITLE
Initial untested buggy implementation

### PR DESCRIPTION
**Pull request by montaguegabe**
I started tweaking the PyTorch with these resources in mind:

If we model our optimizer after LBFGS it reduces a lot of clutter for now.
https://pytorch.org/docs/stable/_modules/torch/optim/lbfgs.html

See here for how to calculate a Hessian-vector product: https://discuss.pytorch.org/t/calculating-hessian-vector-product/11240/3. We can just use this in our code now. I believe whenever MATLAB used the `hess` function we can use this technique.

Also deleted Tensorflow.

Hopefully will finish Steihaug before meeting tomorrow

### CODE REVIEW

File: cifar10-pytorch/tr_cg.py [hunk 1]
```diff
@@ -7,73 +7,35 @@
...
-        # trust_region_param = cg_steihaug()
-        # Give the information to the optimizer via step. Probably by override.
-        # Need to know how to forward with customized weight.
+
+        # Propogate forward to get error
         output = model(data)
 
-        loss = F.cross_entropy(output, target)
+        # Create gradients
+        loss = F.mse_loss(output, target) # this is the loss from MATLAB
```
Comment (ByzanTine) [inline cifar10-pytorch/tr_cg.py]: I think model type is classification, not mse.

File: cifar10-pytorch/tr_cg.py [hunk 2]
```diff
@@ -122,43 +84,35 @@ def main():
...
         default=0.01,
         metavar='LR',
         help='learning rate (default: 0.01)')
-    parser.add_argument(
-        '--momentum',
-        type=float,
-        default=0.5,
-        metavar='M',
-        help='SGD momentum (default: 0.5)')
-    parser.add_argument(
```
Comment (ByzanTine) [inline cifar10-pytorch/tr_cg.py]: Why delete cuda, I think by default this should be trained on GPU.

File: cifar10-pytorch/tr_cg.py (hunk 1 above)
Comment (montaguegabe) [reply cifar10-pytorch/tr_cg.py]: I did this because I believe MATLAB did MSE even though it was classification

File: cifar10-pytorch/tr_cg.py (hunk 2 above)
Comment (montaguegabe) [reply cifar10-pytorch/tr_cg.py]: We should re-add it, but for now it is easier to debug if everything is on the CPU.

File: cifar10-pytorch/tr_cg.py (hunk 1 above)
Comment (montaguegabe) [reply cifar10-pytorch/tr_cg.py]: Actually, I didn't realize before it seems like they define the error as cross-entropy error, but they define the loss as MSE+regularization

Comment (ByzanTine) [reply cifar10-pytorch/tr_cg.py]: Let's put a note here annotating the loss, since MSE for softmax layer is totally against the general practice. 

I would approve for the sake of reproducing the original result.

File: cifar10-pytorch/tr_cg.py [hunk 3]
```diff
@@ -7,73 +7,39 @@
 from optimizer_tr_cg import TRCG
 from torchvision import datasets, transforms
 
-
```
Comment (ByzanTine) [inline cifar10-pytorch/tr_cg.py]: While doing pair programming, I think your method is better. 

Great work!

File: cifar10-pytorch/optimizer_tr_cg.py [hunk 4]
```diff
@@ -1,119 +1,241 @@
...
+        for p in self._params:
+            numel = p.numel()
+            # view as to avoid deprecated pointwise semantics
+            p.data.add_(step_size, update[offset:offset + numel].view_as(p.data))
+            offset += numel
+        assert offset == self._numel()
+
+
+    # This is based on MATLAB and is not optimized for speed
+    def solve_tr_subprob(self, hess_fn, grad, delta, error_tol, max_iters, x0):
```
Comment (ByzanTine) [inline cifar10-pytorch/optimizer_tr_cg.py]: Let's pull this function out to be some tr_steihug.py or something (This could also be done in another PR though). 
And add unit-tests (sanity check) for in a later PR.

Comment (montaguegabe) [reply cifar10-pytorch/optimizer_tr_cg.py]: Yes we will need lots of unit tests as there are likely a few mistakes in here. Will split out into another file

**Comment by ByzanTine** [comment]
Please branch out the two new features(commits) into a different branch and merge the steihug stuff first.

**Comment by ByzanTine** [comment]
Code runs me! Great:) 

But I want to put it strictly, for a better collaboration, please branch out the non-trivial add-on features. 
 
I think even when the code runs, the added-on features (untested full gradient hessian) should not be committed in the same PR. This is making it really hard to collaborate/review/justify.

**Comment by montaguegabe** [comment]
Ah ok sorry yeah I got carried away a bit. I understand how I made it difficult to collaborate there. I was just worried about merging a PR that has so much code that doesn't even run. I'll separate the full-hessian into a separate PR. There are a lot of bugs in it that I fixed with the latest 2 commits though

**Comment by ByzanTine** [comment]
Yes, sure, at the mean time I will submit merge request to your development branch then...

**Comment by ByzanTine** [comment]
Let's actually merge this anyway for the ease of development...
